File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter13/XYSliderDemo/XYSliderDemo/XYSlider.cpp
//
// XYSlider.xaml.cpp
// Implementation of the XYSlider class
//

#include "pch.h"
#include "XYSlider.h"

using namespace XYSliderDemo;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Interop;

DependencyProperty^ XYSlider::valueProperty = 
    DependencyProperty::Register("Point", 
                                 TypeName(Point::typeid), 
                                 TypeName(XYSlider::typeid), 
                                 ref new PropertyMetadata(Point(0.5f, 0.5f), 
                                 ref new PropertyChangedCallback(&XYSlider::OnDependencyPropertyValueChanged)));

XYSlider::XYSlider()
{
    this->DefaultStyleKey = "XYSliderDemo.XYSlider";
}

DependencyProperty^ XYSlider::ValueProperty::get()
{
    return valueProperty;
}

Point XYSlider::Value::get()
{
    return (Point)GetValue(ValueProperty);
}

void XYSlider::Value::set(Point value)
{
    SetValue(ValueProperty, value);
}

void XYSlider::OnApplyTemplate()
{
    // Detach event handlers
    if (contentPresenter != nullptr)
    {
        contentPresenter->ManipulationStarted -= contentPresenterManipulationStartedEventToken;
        contentPresenter->ManipulationDelta -= contentPresenterManipulationDeltaEventToken;
        contentPresenter->SizeChanged -= contentPresenterSizeChangedEventToken;
    }

    // Get new parts
    crossHairPart = dynamic_cast<FrameworkElement^>(GetTemplateChild("CrossHairPart"));
    contentPresenter = dynamic_cast<ContentPresenter^>(GetTemplateChild("ContentPresenterPart"));

    // Attach event handlers
    if (contentPresenter != nullptr)
    {
        contentPresenter->ManipulationMode = ManipulationModes::TranslateX | ManipulationModes::TranslateY;
        contentPresenterManipulationStartedEventToken = contentPresenter->ManipulationStarted += ref new ManipulationStartedEventHandler(this, &XYSlider::OnContentPresenterManipulationStarted);
        contentPresenterManipulationDeltaEventToken = contentPresenter->ManipulationDelta += ref new ManipulationDeltaEventHandler(this, &XYSlider::OnContentPresenterManipulationDelta);
        contentPresenterSizeChangedEventToken = contentPresenter->SizeChanged += ref new SizeChangedEventHandler(this, &XYSlider::OnContentPresenterSizeChanged);
    }

    // Make cross-hair transparent to touch
    if (crossHairPart != nullptr)
    {
        crossHairPart->IsHitTestVisible = false;
    }
}

void XYSlider::OnContentPresenterManipulationStarted(Object^ sender, ManipulationStartedRoutedEventArgs^ args)
{
    RecalculateValue(args->Position);
}

void XYSlider::OnContentPresenterManipulationDelta(Object^ sender, ManipulationDeltaRoutedEventArgs^ args)
{
    RecalculateValue(args->Position);
}

void XYSlider::OnContentPresenterSizeChanged(Object^ sender, SizeChangedEventArgs^ args)
{
    SetCrossHair();
}

void XYSlider::RecalculateValue(Point absolutePoint)
{
    double x = max(0.0, min(1.0, absolutePoint.X / contentPresenter->ActualWidth));
    double y = max(0.0, min(1.0, absolutePoint.Y / contentPresenter->ActualHeight));
    this->Value = Point((float)x, (float)y);
}

void XYSlider::SetCrossHair()
{
    if (contentPresenter != nullptr && crossHairPart != nullptr)
    {
        Canvas::SetLeft(crossHairPart, this->Value.X * contentPresenter->ActualWidth);
        Canvas::SetTop(crossHairPart, this->Value.Y * contentPresenter->ActualHeight);
    }
}

void XYSlider::OnDependencyPropertyValueChanged(Windows::UI::Xaml::DependencyObject^ obj, Windows::UI::Xaml::DependencyPropertyChangedEventArgs^ args)
{
    dynamic_cast<XYSlider^>(obj)->SetCrossHair();
    dynamic_cast<XYSlider^>(obj)->OnValueChanged((Point)args->NewValue);
}

void XYSlider::OnValueChanged(Point value)
{
    ValueChanged(this, value);
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter13/XYSliderDemo/XYSliderDemo/XYSlider.cs
using System;
using Windows.Foundation;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Input;

namespace XYSliderDemo
{
    public class XYSlider : ContentControl
    {
        ContentPresenter contentPresenter;
        FrameworkElement crossHairPart;

        static readonly DependencyProperty valueProperty = 
                DependencyProperty.Register("Value",
                        typeof(Point), typeof(XYSlider), 
                        new PropertyMetadata(new Point(0.5, 0.5), OnValueChanged));

        public event EventHandler<Point> ValueChanged;

        public XYSlider()
        {
            this.DefaultStyleKey = typeof(XYSlider);
        }

        public static DependencyProperty ValueProperty
        {
            get { return valueProperty; }
        }

        public Point Value
        {
            set { SetValue(ValueProperty, value); }
            get { return (Point)GetValue(ValueProperty); }
        }

        protected override void OnApplyTemplate()
        {
            // Detach event handlers
            if (contentPresenter != null)
            {
                contentPresenter.ManipulationStarted -= OnContentPresenterManipulationStarted;
                contentPresenter.ManipulationDelta -= OnContentPresenterManipulationDelta;
                contentPresenter.SizeChanged -= OnContentPresenterSizeChanged;
            }

            // Get new parts
            crossHairPart = GetTemplateChild("CrossHairPart") as FrameworkElement;
            contentPresenter = GetTemplateChild("ContentPresenterPart") as ContentPresenter;

            // Attach event handlers
            if (contentPresenter != null)
            {
                contentPresenter.ManipulationMode = ManipulationModes.TranslateX |
                                                    ManipulationModes.TranslateY;
                contentPresenter.ManipulationStarted += OnContentPresenterManipulationStarted;
                contentPresenter.ManipulationDelta += OnContentPresenterManipulationDelta;
                contentPresenter.SizeChanged += OnContentPresenterSizeChanged;
            }

            // Make cross-hair transparent to touch
            if (crossHairPart != null)
            {
                crossHairPart.IsHitTestVisible = false;
            }

            base.OnApplyTemplate();
        }

        void OnContentPresenterManipulationStarted(object sender, ManipulationStartedRoutedEventArgs args)
        {
            RecalculateValue(args.Position);
        }

        void OnContentPresenterManipulationDelta(object sender, ManipulationDeltaRoutedEventArgs args)
        {
            RecalculateValue(args.Position);
        }

        void OnContentPresenterSizeChanged(object sender, SizeChangedEventArgs args)
        {
            SetCrossHair();
        }

        void RecalculateValue(Point absolutePoint)
        {
            double x = Math.Max(0, Math.Min(1, absolutePoint.X / contentPresenter.ActualWidth));
            double y = Math.Max(0, Math.Min(1, absolutePoint.Y / contentPresenter.ActualHeight));
            this.Value = new Point(x, y);
        }

        void SetCrossHair()
        {
            if (contentPresenter != null && crossHairPart != null)
            {
                Canvas.SetLeft(crossHairPart, this.Value.X * contentPresenter.ActualWidth);
                Canvas.SetTop(crossHairPart, this.Value.Y * contentPresenter.ActualHeight);
            }
        }

        static void OnValueChanged(DependencyObject obj, DependencyPropertyChangedEventArgs args)
        {
            (obj as XYSlider).SetCrossHair();
            (obj as XYSlider).OnValueChanged((Point)args.NewValue);
        }

        protected void OnValueChanged(Point value)
        {
            if (ValueChanged != null)
                ValueChanged(this, value);
        }
    }
}
